custom/zipline/tasks.py

```diff
@@ -0,0 +1,168 @@
+import json
+import requests
+from celery.task import task
+from custom.zipline.api import ProductQuantity, get_order_update_critical_section_key
+from custom.zipline.models import EmergencyOrder, EmergencyOrderStatusUpdate
+from django.conf import settings
+from dimagi.utils.couch import CriticalSection
+from dimagi.utils.logging import notify_exception
+from dimagi.utils.parsing import json_format_datetime
+from requests.auth import HTTPBasicAuth
+
+
+ZIPLINE_STATUS_RECEIVED = 'received'
+ZIPLINE_STATUS_REJECTED = 'rejected'
+ZIPLINE_STATUS_ERROR = 'error'
+
+# The timeout (in seconds) to use when making requests to Zipline
+REQUEST_TIMEOUT = 120
+
+# See send_emergency_order_request()
+RETRY_INTERVAL = 5
+MAX_ATTEMPTS = 3
+
+
+@task(ignore_result=True)
+def send_emergency_order_request(order_id, attempt=1):
+    try:
+        with CriticalSection(
+            [get_order_update_critical_section_key(order_id)],
+            timeout=(REQUEST_TIMEOUT + 10)
+        ):
+            order = EmergencyOrder.objects.get(pk=order_id)
+            order.status = _send_emergency_order_request(order, attempt)
+            order.save()
+    except Exception as e:
+        notify_exception(
+            None,
+            message='[ZIPLINE] Error while sending order',
+            details={
+                'order_id': order_id,
+                'attempt': attempt,
+            }
+        )
+        create_error_record(order, 'Internal error: {}'.format(str(e)))
+        handle_emergency_order_request_retry(order, attempt)
+
+
+def _send_emergency_order_request(order, attempt):
+    """
+    Sends the emergency order request to Zipline.
+    :param order: the EmergencyOrder that should be sent
+    :param attempt: the current attempt number; in the event of errors, a total of MAX_ATTEMPTS will
+    be made, separated by a wait time of RETRY_INTERVAL minutes
+    :return: the new status to be set on the order
+    """
+    order.zipline_request_attempts += 1
+
+    json_payload = get_json_payload_from_order(order)
+    json_payload = json.dumps(json_payload)
+
+    response = requests.post(
+        settings.ZIPLINE_API_URL,
+        auth=HTTPBasicAuth(settings.ZIPLINE_API_USER, settings.ZIPLINE_API_PASSWORD),
+        data=json_payload,
+        headers={'Content-Type': 'application/json'},
+        timeout=REQUEST_TIMEOUT
+    )
+
+    if response.status_code != 200:
+        handle_emergency_order_request_retry(order, attempt)
+        create_error_record(order, 'Received HTTP Response {} from Zipline'.format(response.status_code))
+        return EmergencyOrderStatusUpdate.STATUS_ERROR
+
+    try:
+        response_json = response.json()
+    except:
```

Comment:
Not sure if `response.json()` ever throws other types of errors, but might want to catch a more specific exception.